### ISSUE TITLE
write() doesn't obey context

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10244]
Migrated from [rt.perl.org#73690](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=73690) (status was 'resolved')

Searchable as RT73690$

**Comment by p5pRT** [comment]
### From zefram@fysh.org

#### Created by zefram@fysh.org



$ cat t0\.pl
format STDOUT =
@&#8203;\<\< @&#8203;\<\<
11\, 22
\.
@&#8203;x = \("A"\, scalar\(write\)\, "Z"\);
print join\("\,"\,@&#8203;x\)\, "\\n";
$ perl t0\.pl
11  22
A\,1\,1\,Z
$

The scalar\(write\) expression is resulting in two list elements\.  There are
many variations on this&#8203;: the stack gets broken if write\(\) is used in
an arithmetic expression\, for example\.  If scalar\(write\) is changed to
scalar\(write\(STDOUT\)\)\, it results in \*three\* list elements&#8203;: \*STDOUT\, 1\, 1\.
In both cases\, the list returned in scalar context is identical to the
list that is returned in list context \(where it's allowed to be a list\)\.

If an explicit return is performed inside the format\, write\(\) returns a
single value \(regardless of context\)\.  write\(STDOUT\) still gets \*STDOUT
tacked on in front of the single return value\.

The behaviour is the same on all Perl versions that I've tried\, from
5\.005 up to 5\.11\.5\.

I believe the presence of the GV in the list is due to a missing PUTBACK
in pp\_enterwrite\(\)\, and that the extra return value when leaving the
format normally is due to pp\_leavewrite\(\) not properly adjusting the
stack when popping the context frame\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.10.0:

Configured by Debian Project at Fri Aug 28 22:30:10 UTC 2009.

Summary of my perl5 (revision 5 version 10 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.6.26-2-amd64, archname=i486-linux-gnu-thread-multi
    uname='linux puccini 2.6.26-2-amd64 #1 smp fri aug 14 07:12:04 utc 2009 i686 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=i486-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.10 -Darchlib=/usr/lib/perl/5.10 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.10.0 -Dsitearch=/usr/local/lib/perl/5.10.0 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Ud_ualarm -Uusesfio -Uusenm -DDEBUGGING=-g -Doptimize=-O2 -Duseshrplib -Dlibperl=libperl.so.5.10.0 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='4.3.2', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /usr/lib64
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.7.so, so=so, useshrplib=true, libperl=libperl.so.5.10.0
    gnulibc_version='2.7'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -L/usr/local/lib'

Locally applied patches:
    


@INC for perl 5.10.0:
    /etc/perl
    /usr/local/lib/perl/5.10.0
    /usr/local/share/perl/5.10.0
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.10
    /usr/share/perl/5.10
    /usr/local/lib/site_perl
    .


Environment for perl 5.10.0:
    HOME=/home/zefram
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/zefram/pub/i686-pc-linux-gnu/bin:/home/zefram/pub/common/bin:/usr/bin:/usr/X11R6/bin:/bin:/usr/local/bin:/usr/games:/home/zefram/usr/perl/util
    PERL_BADLANG (unset)
    SHELL=/usr/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From @doy

Fixed in 8e4ecf2\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Fixed in 8e4ecf2\.

**Comment by p5pRT** [comment]
@doy - Status changed from 'new' to 'resolved'
